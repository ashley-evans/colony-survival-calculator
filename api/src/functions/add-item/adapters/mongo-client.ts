import { MongoClient } from "mongodb";

const mongoURI = process.env["MONGO_DB_URI"];
if (!mongoURI) {
    throw new Error("Misconfigured: URI for MongoDB not provided");
}

const accessKeyID = process.env["AWS_ACCESS_KEY_ID"];
if (!accessKeyID) {
    throw new Error("Misconfigured: AWS Access Key ID not provided");
}

const secretAccessKey = process.env["AWS_SECRET_ACCESS_KEY"];
if (!secretAccessKey) {
    throw new Error("Misconfigured: AWS Secret Access Key not provided");
}

const client = MongoClient.connect(mongoURI, {
    ...(process.env["TEST_ENV"]
        ? {}
        : {
              auth: {
                  username: accessKeyID,
                  password: secretAccessKey,
              },
              authSource: "$external",
              authMechanism: "MONGODB-AWS",
          }),
});

export default client;
