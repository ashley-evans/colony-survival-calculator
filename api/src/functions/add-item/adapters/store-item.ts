import client from "@colony-survival-calculator/mongodb-client";

import type { StoreItemPort } from "../interfaces/store-item-port";

const databaseName = process.env["DATABASE_NAME"];
if (!databaseName) {
    throw new Error("Misconfigured: Database name not provided");
}

const itemCollectionName = process.env["ITEM_COLLECTION_NAME"];
if (!itemCollectionName) {
    throw new Error("Misconfigured: Item collection name not provided");
}

const storeItem: StoreItemPort = async (items) => {
    if (items.length === 0) {
        return true;
    }

    const connection = await client;
    const db = connection.db(databaseName);
    const collection = db.collection(itemCollectionName);

    await collection.deleteMany();
    await collection.insertMany(items);

    return true;
};

export { storeItem };
