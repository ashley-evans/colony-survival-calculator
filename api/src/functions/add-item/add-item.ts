import type { GraphQLEventHandler } from "../../interfaces/GraphQLEventHandler";
import type { Item, ItemInput } from "../../graphql/schema";

const handler: GraphQLEventHandler<ItemInput, Item> = async (event) => {
    console.dir(event);
    return {
        id: event.arguments.name,
    };
};

export { handler };
