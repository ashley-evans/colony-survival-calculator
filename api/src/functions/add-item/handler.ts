import dayjs from "dayjs";

import type { GraphQLEventHandler } from "../../interfaces/GraphQLEventHandler";
import type { Item, ItemInput } from "../../graphql/schema";

const handler: GraphQLEventHandler<ItemInput, Item> = async (event) => {
    console.log(`The date is: ${dayjs().format()}`);
    console.dir(event);
    return {
        id: event.arguments.name,
    };
};

export { handler };
