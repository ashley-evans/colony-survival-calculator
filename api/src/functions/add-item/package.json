{
    "name": "@colony-survival-calculator/add-item-function",
    "private": true,
    "version": "0.0.0",
    "dependencies": {
        "@aws-sdk/client-s3": "3.289.0",
        "@colony-survival-calculator/mongodb-client": "file:../../common/mongodb",
        "ajv": "8.12.0",
        "aws4": "1.12.0",
        "mongodb": "5.1.0"
    }
}
