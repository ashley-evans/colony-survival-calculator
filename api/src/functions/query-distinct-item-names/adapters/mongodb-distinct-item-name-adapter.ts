import client from "@colony-survival-calculator/mongodb-client";

import { ItemDatabasePort } from "../interfaces/item-database-port";

const databaseName = process.env["DATABASE_NAME"];
if (!databaseName) {
    throw new Error("Misconfigured: Database name not provided");
}

const itemCollectionName = process.env["ITEM_COLLECTION_NAME"];
if (!itemCollectionName) {
    throw new Error("Misconfigured: Item collection name not provided");
}

const queryDistinctItemNames: ItemDatabasePort = async () => {
    const db = (await client).db(databaseName);
    const collection = db.collection(itemCollectionName);

    return collection.distinct("name");
};

export { queryDistinctItemNames };
