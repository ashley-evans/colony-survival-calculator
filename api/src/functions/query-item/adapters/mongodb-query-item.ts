import client from "@colony-survival-calculator/mongodb-client";
import { Document, Filter } from "mongodb";

import type { Item } from "../../../types";
import type {
    QueryItemByCreatorCountSecondaryPort,
    QueryItemByFieldSecondaryPort,
} from "../interfaces/query-item-secondary-port";

const databaseName = process.env["DATABASE_NAME"];
if (!databaseName) {
    throw new Error("Misconfigured: Database name not provided");
}

const itemCollectionName = process.env["ITEM_COLLECTION_NAME"];
if (!itemCollectionName) {
    throw new Error("Misconfigured: Item collection name not provided");
}

const queryItemByField: QueryItemByFieldSecondaryPort = async (
    name,
    creator
) => {
    const db = (await client).db(databaseName);
    const collection = db.collection<Item>(itemCollectionName);
    const filter: Filter<Item> = {
        ...(name ? { name } : {}),
        ...(creator ? { creator } : {}),
    };

    return collection.find(filter).toArray();
};

const queryItemByCreatorCount: QueryItemByCreatorCountSecondaryPort = async (
    minimumCreators,
    name
) => {
    const db = (await client).db(databaseName);
    const collection = db.collection<Item>(itemCollectionName);
    const pipeline: Document[] = [
        {
            $group: {
                _id: "$name",
                recipes: {
                    $addToSet: "$$ROOT",
                },
                recipeCount: {
                    $sum: 1,
                },
            },
        },
        {
            $match: {
                recipeCount: {
                    $gte: minimumCreators,
                },
            },
        },
        {
            $unwind: {
                path: "$recipes",
            },
        },
        {
            $replaceRoot: {
                newRoot: "$recipes",
            },
        },
    ];

    if (name) {
        pipeline.unshift({ $match: { name } });
    }

    const items = collection.aggregate<Item>(pipeline);

    return items.toArray();
};

export { queryItemByField, queryItemByCreatorCount };
