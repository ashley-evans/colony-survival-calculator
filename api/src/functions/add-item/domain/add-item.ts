import Ajv from "ajv";

import ItemsSchema from "../../../json/schemas/items.json";
import type { Items } from "../../../types";
import { storeItem } from "../adapters/store-item";
import type { AddItemPrimaryPort } from "../interfaces/add-item-primary-port";

const ajv = new Ajv();
const validateItems = ajv.compile<Items>(ItemsSchema);

function parseItems(input: string): Items {
    try {
        const parsed = JSON.parse(input);
        if (validateItems(parsed)) {
            return parsed;
        } else {
            throw validateItems.errors;
        }
    } catch (ex) {
        throw `Validation Error: ${ex}`;
    }
}

const addItem: AddItemPrimaryPort = async (items) => {
    const parsedItems = parseItems(items);
    const itemMap = new Set<string>(parsedItems.map((item) => item.name));
    for (const item of parsedItems) {
        for (const requirement of item.requires) {
            if (!itemMap.has(requirement.name)) {
                throw `Missing requirement: ${requirement.name} in ${item.name}`;
            }
        }
    }

    return storeItem(parsedItems);
};

export { addItem };
