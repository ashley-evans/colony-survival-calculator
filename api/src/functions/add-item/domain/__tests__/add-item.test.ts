import { Items, Tools } from "../../../../types";
import { createItem } from "../../../../../test";

import { storeItem } from "../../adapters/store-item";

jest.mock("../../adapters/store-item", () => ({
    storeItem: jest.fn(),
}));

const mockStoreItem = storeItem as jest.Mock;

import { addItem } from "../add-item";
import { Item } from "../../../../graphql/schema";

const validItem = createItem({
    name: "item name 1",
    createTime: 2,
    output: 3,
    requirements: [],
    minimumTool: Tools.none,
    maximumTool: Tools.none,
    width: 10,
    height: 2,
});
const validItemWithReqs = createItem({
    name: "item name 2",
    createTime: 1,
    output: 2,
    requirements: [{ name: "item name 1", amount: 2 }],
    minimumTool: Tools.none,
    maximumTool: Tools.none,
});
const validItems = [validItem, validItemWithReqs];

const errorLogSpy = jest
    .spyOn(console, "error")
    .mockImplementation(() => undefined);

beforeEach(() => {
    mockStoreItem.mockReset();
    mockStoreItem.mockResolvedValue(true);

    errorLogSpy.mockClear();
});

describe.each([
    ["invalid JSON", "{ invalid JSON"],
    ["a non JSON array", JSON.stringify({})],
    [
        "an item with a missing name",
        JSON.stringify([
            {
                createTime: 2,
                output: 1,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with missing creation time",
        JSON.stringify([
            {
                name: "test",
                output: 1,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with non-numeric creation time",
        JSON.stringify([
            {
                name: "test",
                createTime: "test",
                output: 1,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with negative creation time",
        JSON.stringify([
            {
                name: "test",
                createTime: -1,
                output: 1,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with missing output amount",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with non-numeric output amount",
        JSON.stringify([
            {
                name: "test",
                createTime: 1,
                output: "test",
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with negative output amount",
        JSON.stringify([
            {
                name: "test",
                createTime: 1,
                output: -1,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with missing requirements array",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with non-array requirements",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: "test",
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with a non-object requirement",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: ["test"],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with a requirement that is missing a name",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [{ amount: 1 }],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with a requirement that is missing requirement amount",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [{ name: "wibble" }],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
            {
                name: "wibble",
                createTime: 1,
                output: 1,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with a requirement that has a non-numeric requirement amount",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [{ name: "wibble", amount: "test" }],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
            {
                name: "wibble",
                createTime: 1,
                output: 1,
                requires: [],
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with non-object farm size",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: "wibble",
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with farm size that is missing a width",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    height: 1,
                },
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with farm size that has a non-numeric width",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    width: "test",
                    height: 1,
                },
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with farm size that is missing a height",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    width: 1,
                },
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with farm size that has a non-numeric height",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    width: 1,
                    height: "test",
                },
                minimumTool: Tools.none,
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with a missing minimum tool",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    width: 1,
                    height: 1,
                },
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with an unknown minimum tool",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    width: 1,
                    height: 1,
                },
                minimumTool: "unknown",
                maximumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with a missing maximum tool",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    width: 1,
                    height: 1,
                },
                minimumTool: Tools.none,
            },
        ]),
    ],
    [
        "an item with an unknown maximum tool",
        JSON.stringify([
            {
                name: "test",
                createTime: 2,
                output: 1,
                requires: [],
                size: {
                    width: 1,
                    height: 1,
                },
                minimumTool: Tools.none,
                maximumTool: "unknown",
            },
        ]),
    ],
])(
    "handles invalid input (schema validation) given %s",
    (_: string, input: string) => {
        test("does not store any items in database", async () => {
            try {
                await addItem(input);
            } catch {
                // Ignore
            }

            expect(storeItem).not.toHaveBeenCalled();
        });

        test("throws a validation error", async () => {
            const expectedError = "Validation Error:";

            expect.assertions(1);
            await expect(addItem(input)).rejects.toEqual(
                expect.stringContaining(expectedError)
            );
        });
    }
);

describe.each([
    [
        "unknown item requirements",
        validItemWithReqs,
        "Missing requirement: item name 1 in item name 2",
    ],
    [
        "invalid min/max tool combination (none above stone)",
        createItem({
            name: "test item",
            createTime: 1,
            output: 2,
            requirements: [],
            minimumTool: Tools.stone,
            maximumTool: Tools.none,
        }),
        "Invalid item: test item, minimum tool is better than maximum tool",
    ],
    [
        "invalid min/max tool combination (stone above copper)",
        createItem({
            name: "test item",
            createTime: 1,
            output: 2,
            requirements: [],
            minimumTool: Tools.copper,
            maximumTool: Tools.stone,
        }),
        "Invalid item: test item, minimum tool is better than maximum tool",
    ],
    [
        "invalid min/max tool combination (copper above iron)",
        createItem({
            name: "test item",
            createTime: 1,
            output: 2,
            requirements: [],
            minimumTool: Tools.iron,
            maximumTool: Tools.copper,
        }),
        "Invalid item: test item, minimum tool is better than maximum tool",
    ],
    [
        "invalid min/max tool combination (iron above bronze)",
        createItem({
            name: "test item",
            createTime: 1,
            output: 2,
            requirements: [],
            minimumTool: Tools.bronze,
            maximumTool: Tools.iron,
        }),
        "Invalid item: test item, minimum tool is better than maximum tool",
    ],
    [
        "invalid min/max tool combination (bronze above steel)",
        createItem({
            name: "test item",
            createTime: 1,
            output: 2,
            requirements: [],
            minimumTool: Tools.steel,
            maximumTool: Tools.bronze,
        }),
        "Invalid item: test item, minimum tool is better than maximum tool",
    ],
])("handles item with %s", (_: string, item: Item, expectedError: string) => {
    const input = JSON.stringify([item]);

    test("does not store any items in database", async () => {
        try {
            await addItem(input);
        } catch {
            // Ignore
        }

        expect(storeItem).not.toHaveBeenCalled();
    });

    test("throws a validation error", async () => {
        expect.assertions(1);
        await expect(addItem(input)).rejects.toThrowError(expectedError);
    });
});

describe.each([
    ["a single item", [validItem]],
    ["multiple items", validItems],
])("handles valid input with %s", (_: string, expected: Items) => {
    const input = JSON.stringify(expected);

    test("stores provided item in database", async () => {
        await addItem(input);

        expect(mockStoreItem).toHaveBeenCalledTimes(1);
        expect(mockStoreItem).toHaveBeenCalledWith(expected);
    });

    test("returns success", async () => {
        const actual = await addItem(input);

        expect(actual).toBe(true);
    });
});

describe("error handling", () => {
    test("returns failure if the storage of items failed", async () => {
        const input = JSON.stringify(validItems);
        mockStoreItem.mockResolvedValue(false);

        const actual = await addItem(input);

        expect(actual).toBe(false);
    });

    test("throws an error if an unhandled exception is thrown while storing items", async () => {
        const input = JSON.stringify(validItems);
        const expectedError = new Error("test error");
        mockStoreItem.mockRejectedValue(expectedError);

        expect.assertions(1);
        await expect(addItem(input)).rejects.toEqual(expectedError);
    });

    test("logs the error message to console if an unhandled exception occurs while storing items", async () => {
        const input = JSON.stringify(validItems);
        const expectedError = new Error("test error");
        mockStoreItem.mockRejectedValue(expectedError);

        try {
            await addItem(input);
        } catch {
            // Ignore
        }

        expect(errorLogSpy).toHaveBeenCalledTimes(1);
        expect(errorLogSpy).toHaveBeenCalledWith(expectedError);
    });
});
