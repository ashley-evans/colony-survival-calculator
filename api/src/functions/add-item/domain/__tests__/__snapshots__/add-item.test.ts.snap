// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles invalid input (schema validation) given a non JSON array throws a validation error 1`] = `"Validation Error: [{"instancePath":"","schemaPath":"#/type","keyword":"type","params":{"type":"array"},"message":"must be array"}]"`;

exports[`handles invalid input (schema validation) given an item with a invalid creator name throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/creator","schemaPath":"#/items/properties/creator/type","keyword":"type","params":{"type":"string"},"message":"must be string"}]"`;

exports[`handles invalid input (schema validation) given an item with a missing creator name throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0","schemaPath":"#/items/required","keyword":"required","params":{"missingProperty":"creator"},"message":"must have required property 'creator'"}]"`;

exports[`handles invalid input (schema validation) given an item with a missing maximum tool throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0","schemaPath":"#/items/required","keyword":"required","params":{"missingProperty":"maximumTool"},"message":"must have required property 'maximumTool'"}]"`;

exports[`handles invalid input (schema validation) given an item with a missing minimum tool throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0","schemaPath":"#/items/required","keyword":"required","params":{"missingProperty":"minimumTool"},"message":"must have required property 'minimumTool'"}]"`;

exports[`handles invalid input (schema validation) given an item with a missing name throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0","schemaPath":"#/items/required","keyword":"required","params":{"missingProperty":"name"},"message":"must have required property 'name'"}]"`;

exports[`handles invalid input (schema validation) given an item with a non-object requirement throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/requires/0","schemaPath":"#/items/properties/requires/items/type","keyword":"type","params":{"type":"object"},"message":"must be object"}]"`;

exports[`handles invalid input (schema validation) given an item with a requirement that has a non-numeric requirement amount throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/requires/0/amount","schemaPath":"#/items/properties/requires/items/properties/amount/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"`;

exports[`handles invalid input (schema validation) given an item with a requirement that is missing a name throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/requires/0","schemaPath":"#/items/properties/requires/items/required","keyword":"required","params":{"missingProperty":"name"},"message":"must have required property 'name'"}]"`;

exports[`handles invalid input (schema validation) given an item with a requirement that is missing requirement amount throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/requires/0","schemaPath":"#/items/properties/requires/items/required","keyword":"required","params":{"missingProperty":"amount"},"message":"must have required property 'amount'"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that has a non-numeric output amount throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0/amount","schemaPath":"#/items/properties/optionalOutputs/items/properties/amount/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that has a non-numeric output likelihood throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0/likelihood","schemaPath":"#/items/properties/optionalOutputs/items/properties/likelihood/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that has an output amount less than one throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0/amount","schemaPath":"#/items/properties/optionalOutputs/items/properties/amount/minimum","keyword":"minimum","params":{"comparison":">=","limit":1},"message":"must be >= 1"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that has output likelihood equal to zero throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0/likelihood","schemaPath":"#/items/properties/optionalOutputs/items/properties/likelihood/exclusiveMinimum","keyword":"exclusiveMinimum","params":{"comparison":">","limit":0},"message":"must be > 0"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that has output likelihood greater than one throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0/likelihood","schemaPath":"#/items/properties/optionalOutputs/items/properties/likelihood/maximum","keyword":"maximum","params":{"comparison":"<=","limit":1},"message":"must be <= 1"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that has output likelihood less than zero throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0/likelihood","schemaPath":"#/items/properties/optionalOutputs/items/properties/likelihood/exclusiveMinimum","keyword":"exclusiveMinimum","params":{"comparison":">","limit":0},"message":"must be > 0"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that is missing a name throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0","schemaPath":"#/items/properties/optionalOutputs/items/required","keyword":"required","params":{"missingProperty":"name"},"message":"must have required property 'name'"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that is missing an output amount throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0","schemaPath":"#/items/properties/optionalOutputs/items/required","keyword":"required","params":{"missingProperty":"amount"},"message":"must have required property 'amount'"}]"`;

exports[`handles invalid input (schema validation) given an item with an optional output that is missing an output likelihood throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/optionalOutputs/0","schemaPath":"#/items/properties/optionalOutputs/items/required","keyword":"required","params":{"missingProperty":"likelihood"},"message":"must have required property 'likelihood'"}]"`;

exports[`handles invalid input (schema validation) given an item with an unknown maximum tool throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/maximumTool","schemaPath":"#/definitions/tools/enum","keyword":"enum","params":{"allowedValues":["none","stone","copper","iron","bronze","steel"]},"message":"must be equal to one of the allowed values"}]"`;

exports[`handles invalid input (schema validation) given an item with an unknown minimum tool throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/minimumTool","schemaPath":"#/definitions/tools/enum","keyword":"enum","params":{"allowedValues":["none","stone","copper","iron","bronze","steel"]},"message":"must be equal to one of the allowed values"}]"`;

exports[`handles invalid input (schema validation) given an item with farm size that has a non-numeric height throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/size/height","schemaPath":"#/items/properties/size/properties/height/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"`;

exports[`handles invalid input (schema validation) given an item with farm size that has a non-numeric width throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/size/width","schemaPath":"#/items/properties/size/properties/width/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"`;

exports[`handles invalid input (schema validation) given an item with farm size that is missing a height throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/size","schemaPath":"#/items/properties/size/required","keyword":"required","params":{"missingProperty":"height"},"message":"must have required property 'height'"}]"`;

exports[`handles invalid input (schema validation) given an item with farm size that is missing a width throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/size","schemaPath":"#/items/properties/size/required","keyword":"required","params":{"missingProperty":"width"},"message":"must have required property 'width'"}]"`;

exports[`handles invalid input (schema validation) given an item with missing creation time throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0","schemaPath":"#/items/required","keyword":"required","params":{"missingProperty":"createTime"},"message":"must have required property 'createTime'"}]"`;

exports[`handles invalid input (schema validation) given an item with missing output amount throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0","schemaPath":"#/items/required","keyword":"required","params":{"missingProperty":"output"},"message":"must have required property 'output'"}]"`;

exports[`handles invalid input (schema validation) given an item with missing requirements array throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0","schemaPath":"#/items/required","keyword":"required","params":{"missingProperty":"requires"},"message":"must have required property 'requires'"}]"`;

exports[`handles invalid input (schema validation) given an item with negative creation time throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/createTime","schemaPath":"#/items/properties/createTime/minimum","keyword":"minimum","params":{"comparison":">=","limit":0},"message":"must be >= 0"}]"`;

exports[`handles invalid input (schema validation) given an item with negative output amount throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/output","schemaPath":"#/items/properties/output/minimum","keyword":"minimum","params":{"comparison":">=","limit":1},"message":"must be >= 1"}]"`;

exports[`handles invalid input (schema validation) given an item with non-array requirements throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/requires","schemaPath":"#/items/properties/requires/type","keyword":"type","params":{"type":"array"},"message":"must be array"}]"`;

exports[`handles invalid input (schema validation) given an item with non-numeric creation time throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/createTime","schemaPath":"#/items/properties/createTime/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"`;

exports[`handles invalid input (schema validation) given an item with non-numeric output amount throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/output","schemaPath":"#/items/properties/output/type","keyword":"type","params":{"type":"number"},"message":"must be number"}]"`;

exports[`handles invalid input (schema validation) given an item with non-object farm size throws a validation error 1`] = `"Validation Error: [{"instancePath":"/0/size","schemaPath":"#/items/properties/size/type","keyword":"type","params":{"type":"object"},"message":"must be object"}]"`;

exports[`handles invalid input (schema validation) given invalid JSON throws a validation error 1`] = `"Validation Error: Unexpected token i in JSON at position 2"`;
